File: src/testers/proofTester.ts
import type {FflonkProof, Groth16Proof, PlonkProof, PublicSignals} from 'snarkjs';
import * as snarkjs from 'snarkjs';
import {readFileSync} from 'fs';
import type {CircuitSignals} from '../types/';
import type {CircomkitConfig} from '../configs';
import {AssertionError} from 'node:assert';

/** A tester that is able to generate proofs & verify them.
 * Use `expectFail` and `expectPass` to test out evaluations. */
export class ProofTester<IN extends string[] = [], P extends CircomkitConfig['protocol'] = 'groth16'> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public readonly verificationKey: any;

  constructor(
    readonly wasmPath: string,
    readonly pkeyPath: string,
    readonly vkeyPath: string,
    readonly protocol: P
  ) {
    this.verificationKey = JSON.parse(readFileSync(vkeyPath).toString()) as typeof this.verificationKey;
    if (this.verificationKey.protocol !== protocol) {
      throw new Error('Protocol mismatch.');
    }
  }

  /** Generate a proof for the witness computed from the given input signals. */
  async prove(input: CircuitSignals<IN>): Promise<{proof: Groth16Proof; publicSignals: PublicSignals}>;
  async prove(input: CircuitSignals<IN>): Promise<{proof: PlonkProof; publicSignals: PublicSignals}>;
  async prove(input: CircuitSignals<IN>): Promise<{proof: FflonkProof; publicSignals: PublicSignals}>;
  async prove(
    input: CircuitSignals<IN>
  ): Promise<
    | {proof: Groth16Proof; publicSignals: PublicSignals}
    | {proof: PlonkProof; publicSignals: PublicSignals}
    | {proof: FflonkProof; publicSignals: PublicSignals}
  > {
    return snarkjs[this.protocol].fullProve(input, this.wasmPath, this.pkeyPath, undefined);
  }

  /** Returns the verification result of a proof for some public signals. */
  async verify(proof: Groth16Proof, publicSignals: PublicSignals): Promise<boolean>;
  async verify(proof: PlonkProof, publicSignals: PublicSignals): Promise<boolean>;
  async verify(proof: FflonkProof, publicSignals: PublicSignals): Promise<boolean>;
  async verify(proof: Groth16Proof | PlonkProof | FflonkProof, publicSignals: PublicSignals): Promise<boolean> {
    return await snarkjs[this.protocol].verify(
      this.verificationKey,
      publicSignals,
      proof as Groth16Proof & PlonkProof & FflonkProof
    );
  }

  /** Expects a verification to pass for this proof and public signals. */
  async expectPass(proof: Groth16Proof, publicSignals: PublicSignals): Promise<void>;
  async expectPass(proof: PlonkProof, publicSignals: PublicSignals): Promise<void>;
  async expectPass(proof: FflonkProof, publicSignals: PublicSignals): Promise<void>;
  async expectPass(proof: Groth16Proof | PlonkProof | FflonkProof, publicSignals: PublicSignals): Promise<void> {
    const ok = await this.verify(proof as Groth16Proof & PlonkProof & FflonkProof, publicSignals);
    if (!ok) {
      throw new AssertionError({
        message: 'Expected proof to be verified.',
        expected: true,
        actual: false,
      });
    }
  }

  /** Expects a verification to fail for this proof and public signals. */
  async expectFail(proof: Groth16Proof, publicSignals: PublicSignals): Promise<void>;
  async expectFail(proof: PlonkProof, publicSignals: PublicSignals): Promise<void>;
  async expectFail(proof: FflonkProof, publicSignals: PublicSignals): Promise<void>;
  async expectFail(proof: Groth16Proof | PlonkProof | FflonkProof, publicSignals: PublicSignals): Promise<void> {
    const ok = await this.verify(proof as Groth16Proof & PlonkProof & FflonkProof, publicSignals);
    if (ok) {
      throw new AssertionError({
        message: 'Expected proof to be not verified.',
        expected: false,
        actual: true,
      });
    }
  }
}

File: crates/circomkit-test/src/proof_tester.rs
use std::path::PathBuf;

use circomkit_core::enums::Protocol;
use circomkit_core::types::CircuitSignals;
use circomkit_prove::ProofOutput;
use circomkit_prove::ProvingBackend;

use crate::error::TestError;

/// A circuit tester focused on proof-level testing.
pub struct ProofTester {
    backend: Box<dyn ProvingBackend>,
    wasm_path: PathBuf,
    pkey_path: PathBuf,
    /// The parsed verification key.
    pub vkey: serde_json::Value,
    /// The protocol used for proofs.
    pub protocol: Protocol,
}

impl ProofTester {
    pub fn new(
        backend: Box<dyn ProvingBackend>,
        wasm_path: PathBuf,
        pkey_path: PathBuf,
        vkey_path: PathBuf,
        protocol: Protocol,
    ) -> Result<Self, TestError> {
        let vkey_str = std::fs::read_to_string(&vkey_path)?;
        let vkey: serde_json::Value = serde_json::from_str(&vkey_str)?;

        Ok(Self {
            backend,
            wasm_path,
            pkey_path,
            vkey,
            protocol,
        })
    }

    /// Generate a proof from circuit inputs.
    pub fn prove(&self, input: &CircuitSignals) -> Result<ProofOutput, TestError> {
        Ok(self
            .backend
            .full_prove(input, &self.wasm_path, &self.pkey_path)?)
    }

    /// Verify a proof.
    pub fn verify(
        &self,
        proof: &serde_json::Value,
        public_signals: &[String],
    ) -> Result<bool, TestError> {
        Ok(self.backend.verify(&self.vkey, public_signals, proof)?)
    }

    /// Assert that verification passes.
    pub fn expect_pass(
        &self,
        proof: &serde_json::Value,
        public_signals: &[String],
    ) -> Result<(), TestError> {
        if !self.verify(proof, public_signals)? {
            return Err(TestError::VerificationExpectedPass);
        }
        Ok(())
    }

    /// Assert that verification fails.
    pub fn expect_fail(
        &self,
        proof: &serde_json::Value,
        public_signals: &[String],
    ) -> Result<(), TestError> {
        if self.verify(proof, public_signals)? {
            return Err(TestError::VerificationExpectedFail);
        }
        Ok(())
    }
}
